grpc-api/src/main/java/io/deephaven/grpc_api/arrow/FlightServiceGrpcImpl.java

```diff
@@ -0,0 +1,541 @@
+                .append(" body=").append(mi.inputStream == null ? "none" : Integer.toString(mi.inputStream.available()))
+                .endl();
+
+        if (mi.header != null && mi.header.headerType() == MessageHeader.RecordBatch && mi.inputStream == null) {
+            //noinspection UnstableApiUsage
+            mi.inputStream = new LittleEndianDataInputStream(new ByteArrayInputStream(CollectionUtil.ZERO_LENGTH_BYTE_ARRAY));
+        }
+
+        return mi;
+    }
+
+    private static final class MessageInfo {
+        Message header = null;
+        BarragePutMetadata app_metadata = null;
+        Flight.FlightDescriptor descriptor = null;
+        @SuppressWarnings("UnstableApiUsage")
+        LittleEndianDataInputStream inputStream = null;
+    }
+
+    /**
+     * This is a stateful marshaller; a PUT stream begins with its schema.
+     */
+    private static class PutMarshaller extends SingletonLivenessManager implements Closeable {
+        private long nextSeq = 0;
+        private BarrageSourcedTable resultTable;
+        private SessionState.ExportBuilder<Table> resultExport;
+        private final SessionState session;
+        private final StreamObserver<Flight.PutResult> observer;
+        private final TreeMap<Long, MessageInfo> pendingSeq = new TreeMap<>();
+
+        private PutMarshaller(
+                final SessionState session,
+                final StreamObserver<Flight.PutResult> observer) {
+            this.session = session;
+            this.observer = observer;
+            this.session.addOnCloseCallback(this);
+            if (observer instanceof ServerCallStreamObserver) {
+                ((ServerCallStreamObserver<Flight.PutResult>) observer).setOnCancelHandler(this::tryClose);
+            }
+        }
+
+        private ChunkType[] columnChunkTypes;
+        private Class<?>[] columnTypes;
+
+        public void parseNext(final MessageInfo mi) {
+            GrpcUtil.rpcWrapper(log, observer, () -> {
+                synchronized (this) {
+                    if (nextSeq == -1) {
+                        throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT, "already received final app_metadata; cannot apply update");
+                    }
+                    if (mi.app_metadata != null) {
+                        final long sequence = mi.app_metadata.sequence();
+                        if (sequence != nextSeq) {
+                            pendingSeq.put(sequence, mi);
+                            return;
+                        }
+                    }
+                }
+
+                if (mi.descriptor != null) {
+                    if (resultExport != null) {
+                        throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT, "only one descriptor definition allowed");
+                    }
+                    if (mi.descriptor.getType() != Flight.FlightDescriptor.DescriptorType.PATH
```

Comment:
This code is out-of-date. You and I discussed having a `publish` method on the ticket router.